core/node/rpc/sync/dynmsgbuf/dynmsgbuf.go

```diff
@@ -0,0 +1,101 @@
+package dynmsgbuf
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+)
+
+const (
+	// maxBufferSize
+	maxBufferSize = 512
+)
+
+var (
+	// ErrBufferFull is returned when the buffer is full.
+	ErrBufferFull = errors.New("buffer is full")
+
+	// ErrClosed is returned when the buffer is closed.
+	ErrClosed = errors.New("buffer is closed")
+)
+
+// DynamicBuffer is a thread-safe, dynamically resizing message buffer.
+type DynamicBuffer[T any] struct {
+	mu         sync.Mutex
+	buffer     []T
+	signalChan chan struct{} // Signals when new items are added
+	closed     int64
+}
+
+// NewDynamicBuffer initializes a new dynamic buffer.
+func NewDynamicBuffer[T any]() *DynamicBuffer[T] {
+	return &DynamicBuffer[T]{
+		buffer:     make([]T, 0),
+		signalChan: make(chan struct{}, 1),
+	}
+}
+
+// AddMessage adds a new item to the buffer.
+func (db *DynamicBuffer[T]) AddMessage(item T) error {
+	if atomic.LoadInt64(&db.closed) == 1 {
+		return ErrClosed
+	}
+
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	if len(db.buffer) >= maxBufferSize {
+		return ErrBufferFull
+	}
+
+	db.buffer = append(db.buffer, item)
+
+	// Non-blocking signal (only if empty, avoids duplicate wake-ups)
+	select {
+	case db.signalChan <- struct{}{}:
+	default:
+	}
+
+	return nil
+}
+
+// GetBatch retrieves up to `batchSize` messages, removing them from the buffer.
+func (db *DynamicBuffer[T]) GetBatch(batchSize int) []T {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	if len(db.buffer) == 0 {
+		return nil
+	}
+
+	// Determine batch size (avoid slicing out of range)
+	if batchSize > len(db.buffer) {
+		batchSize = len(db.buffer)
+	}
+
+	// Copy batch & remove from buffer
+	batch := make([]T, batchSize)
+	copy(batch, db.buffer[:batchSize])
+	db.buffer = db.buffer[batchSize:]
+
+	return batch
+}
```

Comment:
Doing this append-only read from front approach is gonna trash memory needlessly.
Apart from using circular buffers it's possible to make it efficient by using two buffers like this:

```suggestion
func (db *DynamicBuffer[T]) GetBatch(prev []T) []T {
    // Reset if capacity is unused
    if prev == nil || (cap(prev) > minBufferSize * 2 && len(prev) < cap(prev) / 2){
        prev = make([]T, 0, minBufferSize)
    } else {
        // Deref pointers so they can be gc'ed faster
        var zero T
        for i := range prev {
            prev[i] = zero
        }
        prev = prev[:0]
    }
    
	db.mu.Lock()
	defer db.mu.Unlock()
    ret := db.buffer
    db.buffer = prev
	return ret
}
```